internal/pkg/policy/policy_output.go

```diff
@@ -97,6 +111,65 @@ func (p *Output) prepareElasticsearch(
 		agent.Outputs[p.Name] = output
 	}
 
+	// retire api key of removed remote output
+	var toRetireAPIKeys *model.ToRetireAPIKeyIdsItems
+	var removedOutputName string
+	// find the first output that is removed - supposing one output can be removed at a time
+	for agentOutputName, agentOutput := range agent.Outputs {
+		found := false
+		for outputMapKey := range outputMap {
+			if agentOutputName == outputMapKey {
+				found = true
+				break
+			}
+		}
+		if !found {
+			zlog.Info().Str(logger.APIKeyID, agentOutput.APIKeyID).Str("outputName", agentOutputName).Msg("Output removed, will retire API key")
+			toRetireAPIKeys = &model.ToRetireAPIKeyIdsItems{
+				ID:        agentOutput.APIKeyID,
+				RetiredAt: time.Now().UTC().Format(time.RFC3339),
+				Output:    agentOutputName,
```

Comment:
As a follow up after https://github.com/elastic/fleet-server/pull/3051#issuecomment-1820608162, I was planning to store the policy id and revision here to query it later when invalidating api keys.
I realized this doesn't work in all cases, because it can happen that the agent is reassigned to another policy, so the policy might be different than the remote output's policy.
I also can't directly query the output from kibana SO as fleet-server doesn't have access and it might be already deleted.

Reply:
implemented in this commit: https://github.com/elastic/fleet-server/pull/3051/commits/727daa4e0fe1c4d62bb32286f8b7080efac2aeda

the solution is not perfect as it queries the last 100 policy revisions to find the output config, we should do a follow up to add mapping on `data.outputs` field to be able to filter in ES.

Reply:
@michel-laterman please review again

Reply:
This seems like a good solution to get this out, thanks for all the effort!